wasm/wat/parser.go

```diff
@@ -222,24 +283,126 @@ func (p *moduleParser) parseImportFunc(tok tokenType, tokenBytes []byte, _, _ ui
 			return fmt.Errorf("redundant name: %s", name)
 		}
 		fn.funcName = name
-	case tokenLParen:
-		return errors.New("TODO: handle (type..) and inlined type (param..) (result..)")
-		// * typeuse https://www.w3.org/TR/wasm-core-1/#text-typeuse
-		// * inlined type https://www.w3.org/TR/wasm-core-1/#abbreviations%E2%91%A6
+	case tokenLParen: // start fields, ex. (param or (result
+		p.tokenParser = p.beginField
+		return nil
 	case tokenRParen: // end of this import func
-		// TODO: once we handle import types, this won't always be empty
-		if len(p.module.types) == 0 {
-			p.module.types = append(p.module.types, typeFuncEmpty)
+		if p.currentType == nil {
+			p.currentType = typeFuncEmpty
 		}
-		p.module.importFuncs[len(p.module.importFuncs)-1].typeIndex = 0
-		p.currentField = fieldModuleImport
-		p.tokenParser = p.parseImport
+
+		fn := p.module.importFuncs[len(p.module.importFuncs)-1]
+
+		// search for an existing signature that matches the current type
+		for i, t := range p.module.types {
+			if bytes.Equal(p.currentType.params, t.params) && bytes.Equal(p.currentType.results, t.results) {
```

Comment:
this is where it is helpful you used a type alias @mathetake!

Reply:
🕺